README.md
# Salsa Clave

Salsa Clave allows you to create sophisticated, customized forms hosted in
the [Salsa Labs](https://www.salsalabs.com/) platform without resorting to
custom pages or embedded code -- meaning non-technical staff will be able to
more easily update them within the Salsa interface they already know.

## The purpose of Salsa Clave

Working with the frontend of forms output by the Salsa suite of hosted tools
for advocacy, donations and more is hard. The default web forms often have
usability problems that hinder your supporters, leading to fewer donations
and fewer actions. There's misleading or unnecessary text that's hardcoded
into the form; there's HTML tags with no classes (and sometimes text with no
HTML tags); there are buttons or important information in hard-to-find places.

Salsa Clave is a small set of JavaScript files (primarily
[jQuery](https://jquery.com)) that refines the Salsa forms' frontend.

Its primary purpose is to allow you to reorder, reword and retheme pieces of
the web forms Salsa outputs to suit your needs. Its method for most of this
work is to rewrite output in semantic, classed elements, and to provide the
ability for you to change hardcoded text in forms.

## How to use these files

Salsa Clave has code that's designed to be read. Each script file is heavily
commented, and includes features that you can enable or disable to suit your
needs.

Perhaps most importantly, because Salsa Clave allows you to rewrite the wording
on the forms, you'll want to determine what wording works for you, and set that
in the code.

## Features (version 0.8)

Salsa Clave is modular. The features listed below are optional, and are turned
off by default -- you only use and enable the features you need.

Each script file corresponds to a different package in Salsa's suite.
Currently, this includes:

* **clave-core.js** is required for any use, and is a good place to start
reading
* **clave-advocacy.js (in development)**, for use with petitions and targeted actions
* **clave-donations.js**, for use with the donations package
* **clave-profiles.js**, for use with the profile manager and unsubscribe pages
* **clave-storefronts.js**, for use with the storefronts package

Currently there is not a script file for the events or chapter Salsa packages.
The email package has no frontend display (outside of email) and therefore has
no script file.

### Core

Salsa Clave assigns semantic class names to all fields, allowing you to target
them specifically.

Want to highlight the zip code field on a signup page, the "personal message"
field on an action page, or the preferred donation amount field on your
donation form? By default, many of these fields have no class name, and are
contained in identically-named &lt;div&gt; elements with the class name
"formRow".

Salsa Clave gives them semantic names like "zip", "comment", and
"donation_amount--50". You can then easily target them with CSS rules or
further jQuery manipulation.

Additional features in core you can enable:

* Remove forced clearing (empty &lt;div&gt; elements and &lt;br&gt; tags).

* Add accessibility attributes to required form fields and error messages
(specifically, the HTML5 "required" attribute and the "aria-required"
attribute on required fields, and the WAI-ARIA role "alert" on error messages).

* Set a default country for all country fields.

* Change corresponding address labels based on the country selected -- e.g.
when "United States" is selected, "State" and "Zip code" are labels; when other
countries are selected, these are changed to "Province" and "Postal code".

### Donations

* Remove the "required" notation on donation form fields, since nearly every
field on a donation form is required.

* Move the donation amount fields before the other fields, an emerging best
practice.

* Add semantic classes to each of the donation amounts, so you can target
and style them specifically as necessary.

* Pre-select a donation amount.

* Pre-fill a donation amount based on a variable in a URL. This is particularly
useful for donation incentives (when you may have multiple tiers and want to
link directly to pre-filled donations for them) or in combination with Salsa's
[dynamic merge fields](https://help.salsalabs.com/entries/22535062-Dynamic-Content-and-Merge-Fields-for-Email-Blasts#3),
potentially allowing you to pre-fill a donation form based on a supporter's
previous donation.

* Set HTML5 field types on numerical fields to trigger a number keypad for
mobile users.

* Improve the "in honor of" and "in memory of" sections with true headers,
and optionally make them collapsible.

* Place the "pre-donation text" directly after donation amounts, instead of
directly before.

* Place the terms of service agreement at the end of the form, out of the
way.

### Advocacy

* [In development] Alter the list of petition signatures from a table into
&lt;div&gt; elements that can be floated or positioned (for instance,
responsively).

* [In development] Additional improvements to come.

### Profile manager

* Customize the hard-coded text on your default unsubscribe page.

* Add custom text to each stage of the unsubscribe process (often a four-page
process). Salsa does not provide any other mechanism for changing this text.

* [In development] Customize the hard-coded welcome text that greets
supporters when they load the "profile manager" page (often encountered when
clicking "update your profile" in an email from Salsa). Salsa does not provide
any other mechanism for changing this text.

* [In development] Remove tabs for packages you may not use (like local events)
or want supporters to have access to (like all past donations and purchases).

* [In development] Clarify the meaning for your supporters of "Group
Membership," which corresponds to Salsa's supporter groups and will almost
always be interpreted as "email lists to which I'm subscribed."

### Storefronts

* Add semantic classes to storefront items and their elements to make it easier
to style their display.

* Add classes to storefront pages, item details and checkout pages, which
otherwise can't be targeted.

* Automatically set the page &lt;title&gt; element to match the contents of the
first &lt;h1&gt; element, if one exists.

* Link item titles and images to the detail/purchase page.

* Convert storefront checkout &lt;table&gt; elements to classed &lt;div&gt;s
that won't break layouts or flow off-screen at narrow widths.

## Caveats

Salsa Clave is a set of JavaScript files that act on code output from Salsa.
This means that they can't (and don't) act on the page until the page loads. In
the case of some features, this means elements might "jump around" in front of
your eyes as the page is loaded, the scripts are loaded, and the scripts
perform their functions. This will generally only be noticeable when it
involves moving large parts of a form, and these situations are noted in the
scripts.

This effect can be mitigated to some extent by minifying your files (see
"Installation"). Additionally, for many users the larger form elements that are
being acted on will be "below the fold" -- that is, off the screen below
introductory text or images at the top of your form -- and will thus not be
noticeable.

Finally, Salsa Clave is designed to increase usability and beauty in your Salsa
pages. However, it does not ship with any defined CSS styles -- it simply gives
you the tools to more easily apply your own.

## Installation

Salsa Clave is meant to be customized to your use. That means that its features
are turned off by default, and heavily commented to explain how they work. So
instead of simply grabbing the code and uploading it to your account, you'll
want to enable the features you need, compress the files to strip out the
comments, and then upload that customized and compressed file to your Salsa
account.

Unfortunately, because Salsa exists in a hosted environment, there's no way to
effectively pull in git updates. Instead, when you upload your minified
versions of the scripts, you should add the latest version number from the top
of the script to the filename, e.g. **clave-1.0.js**.

For testing purposes, you may want to use your web browser to download a copy of
a rendered Salsa page (such as a donation page), add references to your local
uncompressed copy of Salsa Clave, and edit the Clave files until you get the
results you want. If you test "live" on the Salsa servers, you'll have to
delete, re-compress, and re-upload Salsa Clave every time you make a change.

### Step-by-step instructions for beginners

1. You'll want a local copy of Salsa Clave so you can customize it to fit your
needs. Click on the "releases" tab at the top of this page. Next to the most
recent release, click the "zip" icon to download the Salsa Clave package. Unzip
it into a folder.

2. Open the **clave-core.js** file and begin reading. (This file is required for
any use of Salsa Clave.) The code is extensively commented, allowing you to
easily turn on or off features by changing individual values. After you've made
decisions about core Salsa Clave functionality, open each of the other files and
decide which, if any, features you'd like to enable. If you don't use all of
Salsa's features -- for instance, if you don't use the storefronts -- you can
simply delete the corresponding Salsa Clave file.

3. Once you've customized Salsa Clave to fit your needs, decide if you want to
have one single Salsa Clave file with customizations for all your different
Salsa pages (donation pages, advocacy pages, storefronts, users' profile
manager, etc.) or if you want to keep them as separate files.

    * Keeping Salsa Clave files separate will mean you only include the core file
and the appropriate feature file in each kind of template (e.g.
**clave-donations.js** on donation page templates). If you decide to enable or
disable features specific to one section, or if you upgrade to a new version of
Salsa Clave, you'll only have to re-upload the file for that one section.
However, more included files means more return trips for a visitor's web
browser, which can slow down their experience. If you plan to use different
settings for the same types of pages, though (e.g. different settings for
different donation pages), you'll have to keep the files separate.

    * Combining all Salsa Clave files into one file will mean when you turn on or
off individual features or upgrade to a new version of Salsa Clave you will have
to re-compress the entire set of files. However, because it will be the same
single file on all of your Salsa pages, it will get cached in browsers and won't
require additional file downloads, thus speeding up the experience for your
visitors. That said, all of your pages of the same type (e.g. all donation
pages) will have to use the same settings.

    * In either case, I highly recommend you keep a local copy of your customized
Salsa Clave so you don't have to start customizing from scratch in the future.

4. Visit [Google's Closure Compiler](http://closure-compiler.appspot.com/home).
This web application will compress Salsa Clave, stripping out all of the
comments that would increase file size. It will also optionally concatenate
(combine) the files into one single file.

5. Open your local copy of **clave-core.js**. Copy the full text of that file.
In Closure Compiler, highlight all of the existing text in the text box on the
left-hand side and delete it. Then, paste the contents of **clave-core.js** into
this text box. If you are combining multiple Salsa Clave files into one file,
repeat this process with each of the other Salsa Clave files ending in **.js**
that you've customized. Be sure the code you copy from the other files comes
after the code from **clave-core.js**.

6. The default settings for Closure Compiler are good, so simply click
"Compile." On the right-hand side, it will display the compiled code. There will
also be a note at the top saying that it can be accessed at **clave-core.js**.
Click the link to that file and save the file to your computer. I recommend
saving it with the version of Salsa Clave in the file name, such as
**clave-1.0.js** or **clave-core-1.0.js**.

7. Log in to Salsa and go to the "Website" section of the administration
interface. On the right-hand side, click "Images and Files" and then upload
your Salsa Clave file(s). Once uploaded, Salsa will tell you the address of
the file(s) you have uploaded. Note the address(es) as you'll need them in the
next step.

8. On the right-hand side, click "Templates" and find the templates to which
Salsa Clave should be added. Edit each template and insert the following code
at the end of the file just before the &lt;/body&gt; tag:

    ```html
<script src=""></script>
```

    Then, paste the address of the file from the previous step in between the
quotes. If you created multiple Salsa Clave files, repeat this step for each of
the files, ensuring the compressed **clave-core.js** file is first.

### Troubleshooting

If you find Salsa Clave isn't running, first look at the source of the Salsa
page in question to ensure that the code calling the script exists at the bottom
of the page just above the &lt;/body&gt; tag. It's easy to get your templates mixed
up!

By default, Salsa includes a (very old) copy of jQuery in its pages. If you have
a heavily-customized page, it's possible this isn't being loaded, or that
multiple versions of jQuery are being loaded and causing conflicts. If you're
not sure whether jQuery is being loaded on a given page, contact Salsa support.
I've tested Salsa Clave extensively to ensure that it works with the old version
of jQuery on Salsa, but if you believe particular functionality isn't working
because the Salsa version of jQuery doesn't support it, please do
[let me know](https://github.com/rootwork).

Finally, try uploading a copy of your uncompressed Salsa Clave file to Salsa and
including that. I haven't found the compression to break Salsa Clave, so if you
do find this to be the case, please [contact me](https://github.com/rootwork).

## The name

Most components related to Salsa are a play on salsa, the food (e.g.
[Salsa recipes](http://www.salsalabs.com/devs/recipes/),
[WordPress Jalapeno](https://wordpress.org/plugins/wp-jalapeno/)). I decided to relate it to
salsa, the music.
[According to Wikipedia](http://en.wikipedia.org/wiki/Salsa_music#Clave):

> The most fundamental rhythmic element in salsa music is a pattern and concept
> known as clave. Clave is a Spanish word meaning 'code,' 'key,' as in key to a
> mystery or puzzle, or 'keystone,' the wedge-shaped stone in the center of an
> arch that ties the other stones together.

The clave (pronounced clah-vay) is, simply, the key to making the Salsa puzzle
more beautiful.

## Credits

Salsa Clave was created by Ivan Boothe of [Rootwork.org](http://rootwork.org).
I am a frontend developer working primarily with nonprofits and social change
organizations. I've been a user of Salsa (formerly Democracy in Action) in one
form or another since 2004.

The features of Salsa Clave were initially driven and funded by the
[Fellowship of Reconciliation](http://forusa.org), an interfaith organization
working for peace, justice and nonviolence since 1915.

## License

Copyright (C) 2014  Ivan Boothe

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

Full license available at
<https://github.com/rootwork/salsa-clave/blob/master/LICENSE.txt> and
<http://www.gnu.org/licenses/>.


CHANGELOG.txt
Salsa Clave 0.8
----------------
- Removed a feature from donation pages that enabled moving the "other" donation
  field above the main donation values. Unfortunately, this introduced a
  critical bug in which Salsa's own scripts would re-select another value when
  the "other" field was edited.

Salsa Clave 0.7
----------------
- Added additional options for customizing the unsubscribe pages.
- Added the ability to move "pre-donation text" from directly before donation
  amounts to directly after.

Salsa Clave 0.6
----------------
- Added the ability for donation pages to grab a variable from the URL and
  pre-fill a donation amount with that variable.

Salsa Clave 0.5
----------------
- Added support for unsubscribe pages.
- Added the ability to add accessibility attributes to required form fields and
  error messages.

Salsa Clave 0.4
----------------
- Adding support for storefront pages, items, and checkout.
- Improved selector peformance.
- Namespaced variables to the Salsa Clave package.

Salsa Clave 0.3
----------------
- Adding support for donation pages.
- Adding textareas to semantic classes (core).

Salsa Clave 0.2
----------------
- Refactoring to use variables instead of commenting to control features.

Salsa Clave 0.1
----------------
- First official commit.
- Support for core functions: semantic classes, removing forced clears.


clave-donations.js
/*
* CLAVE DONATIONS
* Salsa Clave 0.8
* https://github.com/rootwork/salsa-clave
*
* Purpose: Modifications to Salsa donation pages.
*
* See README.md for complete installation instructions. This file must be used
* with clave-core.js.
*/

/*
* USAGE
*
* The features below are intentionally turned off by default.
*
* Each feature is heavily commented to explain its purpose, so you can choose
* whether or not to enable it.
*
* To enable a feature, change its value to "true".
*
* To disable a feature that has been enabled, change its value to "false".
*
* Remember, this file must be used in conjunction with clave-core.js in order
* to properly function.
*/

/*
* 1. REMOVING 'REQUIRED' NOTATION
*
* On a donation form, virtually everything is required, so these asterisks make
* the form look just a tiny bit more intimidating.
*
* Note that this does not remove required indicators (the HTML5 attribute and
* the WAI-ARIA attribute) set in clave-core.js.
*
* Enable this setting to remove the visual required-field notations.
*/

$clave_donations_remove_required = false;

/*
* 2. PLACING DONATION AMOUNTS AT THE BEGINNING
*
* An emerging best practice of donation forms is to have the donation amount
* at the top, before visitors are asked to start filling out their personal
* information.
*
* Even in single column format, Salsa gives the first set of fields the id
* "left_container" and the second set of fields "right_container". In
* two-column format, you may want to apply some CSS styles to these <div>
* elements to position and space them correctly when reordered.
*
* Note that enabling this feature will cause the form to "jump around" as the
* page loads while things are being reordered. This is mitigated if the form
* fields are below the bottom of the browser window, in which case visitors
* won't see it happen before they scroll down. If the movement is visible,
* you'll have to balance the appearance of this movement with the advantages
* to having the donation amounts first.
*
* Enable this setting to place the donation amounts at the beginning of the
* form.
*/

$clave_donations_prepend_amounts = false;

/*
* 3. PLACING TERMS OF SERVICE NOTICE AT THE END
*
* If you use Democracy Engine as your donation processor, Salsa will insert a
* mandatory terms of service to your form. This usually appears above the
* submit button, and if you move the donation amounts to the beginning (above),
* this will end up in the middle of your form.
*
* Enable this setting to move the Democracy Engine TOS notice to the end of the
* form.
*/

$clave_donations_append_tos = false;

/*
* 4. ADD SEMANTIC CLASSES TO DONATION AMOUNTS
*
* Want to pre-select a donation amount? Want to add pop-up text to that amount?
* Want to emphasize amounts with particular images or CSS styles? You'll need
* to take this step first, so you can target those amounts fields specifically,
* because by default they all have the same classes, "formRow" and "amount".
*
* The classes that are added are of the form "donation_amount--x", where "x" is
* the numerical amount. The existing "formRow" and "amount" classes will be left
* in place.
*
* Enable this setting to add semantic classes to the donation amount fields.
*/

$clave_donations_semantic_classes = false;

/*
* 5. PRE-SELECTING A DONATION AMOUNT
*
* Choose any of your existing donation amounts to be pre-selected when a
* visitor loads the page. They can still choose other donation amounts if they
* wish.
*
* Note that the previous section (adding semantic classes to donation amounts)
* must be enabled in order for this to take effect.
*
* The default setting ('') leaves the donation amounts as they are, with none
* pre-selected.
*
* This only allows you to pre-select one donation amount per page. If you want
* the ability to pre-select multiple donation amounts, see the pre-filling
* feature in the next section.
*
* Enter the donation amount you wish to pre-select (e.g. '50' for $50.00).
*/

$clave_donations_preselect_amount = '';

/*
* 5b. Pre-fill the "other" donation amoutn field from a URL variable
*
* Alternatively, you can use a variable in the URL to pre-fill the "other" field
* with a value, optionally hiding the "other" label and move it above the other
* amount options, or optionally hide the other amount options entirely.
*
* When this is enabled and the URL has a pre-fill amount, the class
* "is-prefilled" will be added to the amount textfield to enable you to target
* theming specific to when it's being used (for instance, to make it more
* prominent).
*
* When this is enabled and the URL has no pre-filled amount, none of the
* settings below will be applied. This is by design, so that you can enable
* pre-filling on a donation form that might also get visitors from other
* sources.
*
* If the URL for your donation page is:
* https://org.salsalabs.com/o/123/p/salsa/donation/common/public/?donate_page_KEY=3210
* then you could add to the URL in the following way to automatically pre-fill
* (and optionally put all focus on) a donation of $45:
* https://org.salsalabs.com/o/123/p/salsa/donation/common/public/?donate_page_KEY=3210&amt=45
*
* (If you are using a custom URL for your donation page that doesn't end with the
* donation page key, use "?amt=" instead of "&amt=".)
*
* You can use this with incentive gifts in useful ways. For instance, if you
* tell people gifts over $50 get a free doodad, and you have a picture of said
* doodad in your email to supporters, you can link that picture directly to the
* donation form with $50 pre-filled.
*
* This can also be used with Salsa's dynamic merge fields, although it takes a
* little extra work. For instance, Salsa's donation history allows you can take
* a supporter's last donation of $50, increase the amount by 10%, and suggest in
* an email that they donate $55. If you were to append that same dynamic field
* to your donation page's URL with "&amt=" in the email, the link for that
* supporter will take them to your donation page with that suggestion
* pre-filled.
*
* Unfortunately, Salsa's dynamic merge fields can't currently inject their
* calculations into a URL field. Instead, you'll have to export your donations,
* open them in Excel, create a new column for "suggested donation," run the
* calculation in Excel, create a matching custom field for the suggested
* donation in Salsa, then import the records. From there, you can use Salsa's
* placeholders for custom fields to insert the value.
*
* For instance, if your donation page is:
* https://org.salsalabs.com/o/123/p/salsa/donation/common/public/?donate_page_KEY=3210
* then view the email's source code and change the URL to:
* https://org.salsalabs.com/o/123/p/salsa/donation/common/public/?donate_page_KEY=3210&amt=[[suggesteddonation]]
*
* Where "suggesteddonation" is the API name of your suggested donation field.
*
* If you are using a custom URL for your donation page that doesn't end with the
* donation page key, use "?amt=" instead of "&amt=".
*
* For more information about Salsa's dynamic merge fields, see:
* https://help.salsalabs.com/entries/22535062-Dynamic-Content-and-Merge-Fields-for-Email-Blasts#3
*
* Since this is a core feature of Salsa's donation and email integration, you
* can and should look to Salsa's support for help with this process. And perhaps
* joining me in pressuring them to support this feature natively!
*/

// Enable "other" value pre-fill from URL
$clave_donations_url_prefill = false;

// Remove the word "Other" in the label
$clave_donations_url_prefill_label_hide = false;

// Insert language above the other values like "Or, choose the following:"
$clave_donations_url_prefill_suffix = '';

// Remove the other values entirely
$clave_donations_url_prefill_isolate = false;

/*
* 6. SET HTML5 INPUT FIELD TYPES
*
* For mobile visitors in particular, setting HTML5 types on your input fields
* will speed their donation process, as their keyboard will automatically
* change to the correct type (for instance, to a numerical keyboard when
* entering the phone number).
*
* Enable the settings for the fields to which you want to apply HTML5 types.
* If you don't have a particular field in your form, you should leave its
* setting on "false".
*/

// Email field
$clave_donations_html5_email = false;

// Phone number field
$clave_donations_html5_phone = false;

/*
* 6b. Zip code field: U.S.-only
*
* If you ONLY have transactions from U.S. addresses, setting this field will
* speed those transactions. Non-U.S. addresses, however, have letters as well
* as numbers, so if you accept donations from outside the United States, you
* should not enable this setting.
*
* Enable this setting to give the zip code field a numerical field type and
* ONLY allow numerical postal codes.
*/

$clave_donations_html5_zip = false;

/*
* 6c. Credit card fields
*
* Set the two credit card fields (number and security code) to a numerical
* pattern or a number format.
*
* Options:
*
* '' (no value, default) = Leaves the fields as they are.
*
* 'safe' = These fields will be given a numerical pattern. In many mobile
* devices, but not all, this will trigger a number pad keyboard to appear.
*
* 'edge' = These fields will be given a number format, which will trigger a
* number pad keyboard to appear in all mobile devices. Hower, in older versions
* of Safari desktop and older versions of Safari mobile (e.g. older iPads),
* Safari will auto-insert commas into these numbers, invalidating them when the
* visitor attempts to submit the form. Unfortunately, there's no reliable way
* to disable the auto-insertion of commas in these older browsers. Only set
* 'edge' if you're confident you don't have visitors using these browsers, or
* have ways for correcting it.
*
* Enable this setting with 'safe' or 'edge' to set the credit card fields to a
* numerical pattern or number, respectively.
*/

$clave_donations_html5_cc = '';

/*
* 6d. 'Other' donation amount
*
* Set the "other" donation amount field to a numerical pattern or a number
* format. Similar to the credit card fields (see above), the purpose of this
* change is to cause a number pad keyboard to appear for mobile users.
*
* Options:
*
* '' (no value, default) = Leaves this field as it is.
*
* 'safe' = These fields will be given a numerical pattern. In many mobile
* devices, but not all, this will trigger a number pad keyboard to appear.
*
* 'edge' = These fields will be given a number format, which will trigger a
* number pad keyboard to appear in all mobile devices. Hower, in older versions
* of Safari desktop and older versions of Safari mobile (e.g. older iPads),
* Safari will auto-insert commas into these numbers, invalidating them when the
* visitor attempts to submit the form. Unfortunately, there's no reliable way
* to disable the auto-insertion of commas in these older browsers. Only set
* 'edge' if you're confident you don't have visitors using these browsers, or
* have ways for correcting it.
*
* Enable this setting with 'safe' or 'edge' to set the "other" donation amount
* field to a numerical pattern or number, respectively.
*/

$clave_donations_html5_donation_other = '';

/*
* 7. IMPROVING THE 'IN HONOR OF' AND 'IN MEMORY OF' SECTIONS
*
* By default, this fieldset is poorly formed, with nonsemantic tags meant to
* replicate headers, and few classes. Additionally, the fieldset takes up quite
* a bit of space, despite the fact that many visitors won't be using it.
*
* Enable the settings below to add true headers to these sections, and make the
* sections collapsible.
*/

// Add a true header to the "in honor of" section
$clave_donations_honorof_header = false;

// Add a true header to the "in memory of" section
$clave_donations_memoryof_header = false;

// Make this section collapsible
$clave_donations_honorof_collapsible = false;

/*
* 8. POSITION PRE-DONATION TEXT AFTER DONATION AMOUNTS
*
* Salsa gives you a field for text that appears at the top of the "donation
* amounts" fieldset. This can be helpful as a way to introduce your amounts or
* talk about the impact of each donation level.
*
* However, sometimes you don't want to give the text that much emphasis. For
* instance you might want to have a note directly under the amounts saying "all
* donations over $50 will receive a free t-shirt."
*
* Enable the setting below to move the "pre-donation text" from the beginning of
* the donation amounts fieldset to the end of the donation amounts fieldset.
*/

$clave_donations_pre_post_donation_text = false;


/*
*  *************************************************************************
*  End of user-configurable options. Do not modify anything below this line.
*  *************************************************************************
*/



// Document ready
$(document).ready(function() {

// Grab variables from the URL
function urlval(name) {
  name = RegExp ('[?&]' + name.replace (/([[\]])/, '\\$1') + '=([^&#]*)');
  return (window.location.href.match (name) || ['', ''])[1];
}

// Set pre-selected amount from URL
var amtpreselect = urlval("amt");

/*
* 1. REMOVING 'REQUIRED' NOTATION
*/

if($clave_donations_remove_required) {
  $('.donation .required').remove();
}

/*
* 2. PLACING DONATION AMOUNTS AT THE BEGINNING
*/

if($clave_donations_prepend_amounts) {
  $('.donation #right_container').insertBefore($('.donation #left_container'));
}

/*
* 3. PLACING TERMS OF SERVICE NOTICE AT THE END
*/

if($clave_donations_append_tos) {
  $('.donation #de_compliance').insertAfter($('#submit'));
}

/*
* 4. ADD SEMANTIC CLASSES TO DONATION AMOUNTS
*/

if($clave_donations_semantic_classes) {
  $('.donation #donation_amount > .formRow input[type="radio"]').each(function(){
    var donateamt_value = $(this).attr("value");
    var donateamt = $.trim(donateamt_value);
    $(this).parent('.formRow').addClass('donation_amount--' + donateamt);
    $(this).siblings('label').addClass('donation_amount--' + donateamt + '__label');
  });
  $('.donation #donation_amount .amountother').addClass('donation_amount--other').removeClass('donation_amount--');
  $('.donation #donation_amount .amountother label').addClass('donation_amount--other__label').removeClass('donation_amount--__label');
}

/*
* 5. PRE-SELECTING A DONATION AMOUNT
*/

if($clave_donations_preselect_amount) {
  $('.donation_amount--' + $clave_donations_preselect_amount + ' input').attr('checked','checked');
}

/*
* 5b. Fill 'other' field with amount set in URL
*/

if($clave_donations_url_prefill && amtpreselect) {
  $('.donation .otherRow #otheramt').val(amtpreselect).addClass('is-prefilled');
  $('.donation .otherRow #other').attr('checked','checked');

  if($clave_donations_url_prefill_label_hide) {
    $('.donation .otherRow label').text("$");
  }

  if($clave_donations_url_prefill_suffix) {
    $('.donation .otherRow').append('<p class="donation_amount_suffix--other">' + $clave_donations_url_prefill_suffix + '</p>');
  }

  if($clave_donations_url_prefill_isolate) {
    $('.donation .formRow.amount').hide();
    $('.donation .formRow.otherRow').show();
    $('.donation .otherRow #other').hide();
  }
}

/*
* 6. SET HTML5 INPUT FIELD TYPES
*/

if($clave_donations_html5_email) {
  $('.donation .email input').get(0).type = 'email';
}

if($clave_donations_html5_phone) {
  $('.donation .phone input').get(0).type = 'tel';
}

/*
* 6b. Zip code field
*/

if($clave_donations_html5_zip) {
  $('.donation .zip input').get(0).type = 'number';
  $('.donation .zip input').attr('pattern','[0-9]*');
}

/*
* 6c. Credit card fields
*/

if($clave_donations_html5_cc == 'safe') {
  $('.donation input#cc_number').attr('pattern','[0-9]*');
  $('.donation .cvv2 input').attr('pattern','[0-9]*');
}

if($clave_donations_html5_cc == 'edge') {
  $('.donation input#cc_number').get(0).type = 'number';
  $('.donation .cvv2 input').get(0).type = 'number';
}

/*
* 6d. 'Other' donation amount
*/

if($clave_donations_html5_donation_other == 'safe') {
  $('.donation input#otheramt').attr('pattern','[0-9]*');
}

if($clave_donations_html5_donation_other == 'edge') {
  $('.donation input#otheramt').get(0).type = 'number';
}

/*
* 7. IMPROVING THE 'IN HONOR OF' AND 'IN MEMORY OF' SECTIONS
*/

if($clave_donations_honorof_header) {
  // Add a true header to the "in honor of" section
  $('.donation #honorof p:contains("in honor of")').replaceWith('<h4 class="honorof__header">In honor of...</h4>');
}

if($clave_donations_memoryof_header) {
  // Add a true header to the "in memory of" section
  $('.donation #honorof p:contains("in memory of")').replaceWith('<h4 class="honorof__header">In memory of...</h4>');
}

if($clave_donations_honorof_collapsible) {
  // Grab all of the .formRows
  var divs = $('.donation #honorof .formRow');
  // Wrap each set in a semantic section div
  for(var i = 0; i < divs.length; i+=3) {
    divs.slice(i, i+3).wrapAll('<div class="honorof__section"></div>');
  }
  // Hide each section by default
  $('.donation .honorof__section').css('display','none');
  // Apply the following to both the default and rewritten headers
  var header_sel = $('.donation .honorof__header, .donation #honorof p:contains("in honor of"), .donation #honorof p:contains("in memory of")');
  // Grab the default attributes of links and set them on the header
  $(function(){
    var linkColor = $('a:link').css('color');
    var linkCursor = $('a:link').css('cursor');
    var linkLine = $('a:link').css('text-decoration');
    header_sel.css('color', linkColor).css('cursor', linkCursor).css('text-decoration', linkLine);
  });
  // Accordionize the headers
  header_sel.click(function () {
    $header = $(this);
    // Get the next element
    $content = $header.next();
    // Open up the content needed -- toggle the slide
    // If visible slide up, if not slide down
    $content.slideToggle(500, function () {
    // Execute this after slideToggle is done
    });
  });
}

/*
* 8. POSITION PRE-DONATION TEXT AFTER DONATION AMOUNTS
*/

if($clave_donations_pre_post_donation_text) {
  $('#pre_donation_text').appendTo($('#donation_amount'));
}

}); // End of document ready
